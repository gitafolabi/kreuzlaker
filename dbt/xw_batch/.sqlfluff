[sqlfluff]
templater = dbt
sql_file_exts = .sql
# Unfortunately, there is not yet a dialect for athena/presto
# https://github.com/sqlfluff/sqlfluff/issues/2168
dialect = ansi
output_line_length = 120

# https://docs.sqlfluff.com/en/stable/rules.html
# L034: do not order columns by complexity
# L031: using table aliases is acceptable, but please use sensible ones...
# L044 : don't fail on "SELECT * ..." because of undetermined columns
exclude_rules = L031, L034, L044

[sqlfluff:rules]
tab_space_size = 4
max_line_length = 120
comma_style = trailing


