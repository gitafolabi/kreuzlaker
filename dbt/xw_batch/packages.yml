packages:
  - package: dbt-labs/dbt_utils
    # https://hub.getdbt.com/dbt-labs/dbt_utils/latest/
    version: 0.9.2
  - package: lalalilo/athena_utils
    # https://hub.getdbt.com/lalalilo/athena_utils/latest/
    version: 0.2.1