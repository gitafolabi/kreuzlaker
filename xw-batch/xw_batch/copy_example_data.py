"""
# Copyjob for example data

Adjusted from
https://gitlab.kreuzwerker.de/alican.kapusuz/cop-de-scoofy-alican/-/blob/master/scoofy_project/scoofy_project_stack.py

"""
import aws_cdk
from aws_cdk import (
    aws_events,
    aws_events_targets,
    aws_iam,
    aws_lambda,
    aws_s3,
    lambda_layer_awscli,
)


def add_copy_scoofy_example_data(
    stack: aws_cdk.Stack, raw_data_bucket: aws_s3.Bucket
) -> str:
    """Copies the scoofy example data to the bucket under /raw/scoofy/journeys/"""
    cron_minute = 10
    cron_hour = "*"
    input_bucket = "xw-d13g-scoofy-data-inputs"
    input_bucket_path = "/data/journeys"
    output_bucket_path = "/raw/scoofy/journeys/"

    # Synchronize raw input bucket with duplicated data bucket
    copy_scoofy_example_data_lambda = aws_lambda.Function(
        stack,
        id="copy-scoofy-example-data",
        runtime=aws_lambda.Runtime.PYTHON_3_9,  # type: ignore
        code=aws_lambda.Code.from_asset("lambda/copyjob_for_example_data"),
        handler="copyjob_for_example_data.sync_bucket_uri",
        environment={
            "INPUT_BUCKET_URI": f"s3://{input_bucket}{input_bucket_path}",
            "OUTPUT_BUCKET_URI": f"s3://{raw_data_bucket.bucket_name}{output_bucket_path}",
        },
        timeout=aws_cdk.Duration.minutes(15),
    )

    synchron_lambda_rule = aws_events.Rule(
        stack,
        id="synchron-rule",
        schedule=aws_events.Schedule.cron(
            minute=f"{cron_minute}",
            hour=f"{cron_hour}",
            month="*",
            week_day="*",
            year="*",
        ),
    )
    synchron_lambda_rule.add_target(
        aws_events_targets.LambdaFunction(copy_scoofy_example_data_lambda)
    )
    # AWS CLI will be installed under /opt/awscli/aws
    copy_scoofy_example_data_lambda.add_layers(
        lambda_layer_awscli.AwsCliLayer(stack, "AwsCliLayer")
    )

    raw_data_bucket.grant_read_write(copy_scoofy_example_data_lambda)

    scoofy_example_data_bucket_arn = f"arn:aws:s3:::{input_bucket}"

    allow_read_access_to_scoofy_statement = aws_iam.PolicyStatement(
        effect=aws_iam.Effect.ALLOW,
        actions=[
            "s3:GetObject",
            "s3:ListBucket",
        ],
        resources=[
            scoofy_example_data_bucket_arn + "/*",  # for s3:GetObject
            scoofy_example_data_bucket_arn,  # for s3:ListBucket
        ],
    )

    copy_scoofy_example_data_lambda.add_to_role_policy(
        allow_read_access_to_scoofy_statement
    )

    return output_bucket_path
