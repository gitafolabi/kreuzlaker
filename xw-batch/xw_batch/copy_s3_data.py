"""
# Copyjob from some s3 bucket path to a s3 bucket in our cdk stack

Adjusted from
https://gitlab.kreuzwerker.de/alican.kapusuz/cop-de-scoofy-alican/-/blob/master/scoofy_project/scoofy_project_stack.py

"""
import aws_cdk
from aws_cdk import (
    aws_events,
    aws_events_targets,
    aws_iam,
    aws_lambda,
    aws_s3,
    lambda_layer_awscli,
)
from constructs import Construct


class CopyS3Data(Construct):
    def __init__(
        self,
        scope: Construct,
        id: str,
        *,
        source_bucket_name: str,
        source_bucket_path: str,
        target_bucket: aws_s3.Bucket,
        target_bucket_path: str,
        schedule_cron_minute: str = "10",
        schedule_cron_hour: str = "*",
    ):
        """Copies the s3 data from a source s3 bucket to the target s3 bucket"""
        super().__init__(scope, id)
        self.source_bucket_name = source_bucket_name
        self.source_bucket_path = source_bucket_path
        self.target_bucket = target_bucket
        self.target_bucket_path = target_bucket_path
        self.schedule_cron_minute = schedule_cron_minute
        self.schedule_cron_hour = schedule_cron_hour

        # Synchronize raw input bucket with duplicated data bucket
        self.copy_data_lambda = aws_lambda.Function(
            self,
            id="copy-data-lambda",
            runtime=aws_lambda.Runtime.PYTHON_3_9,  # type: ignore
            code=aws_lambda.Code.from_asset(
                "lambdas/copyjob_for_s3_data",
                exclude=[
                    # Excluded to make repeatable builds in case these files get compiled by tests
                    "__pycache__",
                ],
            ),
            handler="copyjob_for_s3_data.sync_bucket_uri",
            environment={
                "SOURCE_BUCKET_URI": f"s3://{self.source_bucket_name}{self.source_bucket_path}",
                "TARGET_BUCKET_URI": f"s3://{self.target_bucket.bucket_name}{self.target_bucket_path}",
            },
            timeout=aws_cdk.Duration.minutes(15),
        )
        # For some reason, this is needed to resolve the name in the environment variable
        # https://bobbyhadz.com/blog/aws-cdk-dependson-relation
        self.copy_data_lambda.node.add_dependency(self.target_bucket)

        self.synchron_lambda_rule = aws_events.Rule(
            self,
            id="synchron-rule",
            schedule=aws_events.Schedule.cron(
                minute=f"{self.schedule_cron_minute}",
                hour=f"{self.schedule_cron_hour}",
                month="*",
                week_day="*",
                year="*",
            ),
        )
        self.synchron_lambda_rule.add_target(
            aws_events_targets.LambdaFunction(self.copy_data_lambda)
        )
        # AWS CLI will be installed under /opt/awscli/aws
        self.copy_data_lambda.add_layers(
            lambda_layer_awscli.AwsCliLayer(self, "AwsCliLayer")
        )

        self.target_bucket.grant_read_write(self.copy_data_lambda)

        self.source_bucket_arn = f"arn:aws:s3:::{self.source_bucket_name}"

        self.allow_read_access_to_source_bucket_statement = aws_iam.PolicyStatement(
            effect=aws_iam.Effect.ALLOW,
            actions=[
                "s3:GetObject",
                "s3:ListBucket",
            ],
            resources=[
                self.source_bucket_arn + "/*",  # for s3:GetObject
                self.source_bucket_arn,  # for s3:ListBucket
            ],
        )

        self.copy_data_lambda.add_to_role_policy(
            self.allow_read_access_to_source_bucket_statement
        )
