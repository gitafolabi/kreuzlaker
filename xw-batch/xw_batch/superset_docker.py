from copyreg import constructor
from aws_cdk import aws_ecs_patterns as ecs
from aws_cdk.aws_ecs import ContainerImage
from aws_cdk.aws_ec2 import Vpc
from constructs import Construct


class Superset(Construct):
    def __init__(self, scope: Construct, id: str, vpc: Vpc, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        ecs.ApplicationLoadBalancedFargateService(
            self,
            "SuperSetFargateService",
            vpc=vpc,
            task_image_options=ecs.ApplicationLoadBalancedTaskImageOptions(
                image=ContainerImage.from_registry("apache/superset:latest"),
                container_name="Superset",
            ),
        )
