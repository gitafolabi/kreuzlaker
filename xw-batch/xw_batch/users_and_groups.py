"""
Create users and groups for our org
"""
import re
import typing

from aws_cdk import SecretValue, aws_iam
from constructs import Construct

# We use this as initial password which has to be changed on first login.
# TODO: use `secretsmanager.Secret.fromSecretAttributes` to reference a secret in Secrets Manager? Prob not worth it.
INITIAL_PASSWORD = "N0tV3ryS3cr3t!"

# Default policy: be able to change passwords
# Yes, you have to give the user a non-managed policy for that...
# https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_examples_aws_my-sec-creds-self-manage-password-only.html
# https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_passwords_enable-user-change.html

POLICY_DOCUMENT_ALLOW_PASSWORD_CHANGES = aws_iam.PolicyDocument(
    statements=[
        aws_iam.PolicyStatement(
            sid="ViewAccountPasswordRequirements",
            effect=aws_iam.Effect.ALLOW,
            actions=["iam:GetAccountPasswordPolicy"],
            resources=["*"],
        ),
        aws_iam.PolicyStatement(
            sid="ChangeOwnPassword",
            effect=aws_iam.Effect.ALLOW,
            actions=["iam:GetUser", "iam:ChangePassword"],
            resources=["arn:aws:iam::*:user/${aws:username}"],
        ),
    ]
)


def _validate_user_name(user_name: str):
    """Raises if a name for a user or group is not according to AWS spec

    :param: user_name: the name which should conform to the AWS rules
    """

    # Spec according to error message:
    # userName [...] must contain only alphanumeric characters and/or the following: +=,.@_-
    if not re.match(r"^[a-z+=,.@_-]+$", user_name):
        raise RuntimeError(f"Bad user name: {user_name}; Only [a-z+=,.@_-]+ allowed")


def _validate_group_name(group_name: str):
    """Raises if a name for a group is not according to AWS spec

    :param: user_name: the name which should conform to the AWS rules
    """

    # Spec according to https://docs.aws.amazon.com/IAM/latest/UserGuide/id_groups_create.html
    if not re.match(r"^[A-Za-z0-9+=,.@_-]+$", group_name):
        raise RuntimeError(
            f"Bad group name: {group_name}; Only [A-Za-z0-9+=,.@_-]+ allowed"
        )


def _to_id(name: str) -> str:
    """Converts a name lower case with minus between words.

    :param: name: the name to be converted to a construct id
    :return: the id this name gets converted to
    """
    name = re.sub("(.)([A-Z][a-z]+)", r"\1-\2", name)
    return re.sub("([a-z0-9])([A-Z])", r"\1-\2", name).lower()


class OrgUsersAndGroups(Construct):
    """A construct to create groups and users with some defaults

    Users are created with a default password and must change it on first login
    """

    def __init__(self, scope: Construct, id: str):
        super().__init__(scope, id)

        self.policy_allow_password_changes = aws_iam.ManagedPolicy(
            self,
            "policy-allow-password-changes",
            document=POLICY_DOCUMENT_ALLOW_PASSWORD_CHANGES,
            # Do not set to not have problems when deploying any changes to the policy. See best practises for cdk
            # managed_policy_name="AllowPasswordChanges",
            description="Allow password changed for users.",
        )

        # Users
        self.users: typing.Dict[str, aws_iam.User] = {}
        self.groups: typing.Dict[str, aws_iam.Group] = {}

    def add_org_group(self, group_name: str) -> aws_iam.Group:
        """
        Create an IAM group

        :param: group_name: The name if the group
        :return: group: the aws_iam.Group which got created
        """
        _validate_group_name(group_name)
        id_ = _to_id(group_name)
        if id_ in self.groups:
            raise RuntimeError(f"Group '{group_name}' (id={id_})  already exists.")

        group = aws_iam.Group(self, id=id_, group_name=group_name)
        self.groups[id_] = group
        return group

    def get_group(self, group_name: str) -> aws_iam.Group:
        """
        Return the already created Group for the name

        :param: group_name: the name of the group
        :return: the group if it was already created or raise error
        """
        key = _to_id(group_name)
        if key not in self.groups:
            raise KeyError(f"Group with name {group_name} (id={key}) does not exist.")
        return self.groups[key]

    def add_org_user(
        self,
        user_name: str,
        *,
        groups: typing.Union[str, typing.List[str]] = None,
    ) -> aws_iam.User:
        """Create an IAM user who must change their password immediately.

        :param: groups: The group(s) to which this user should belong
        :param: user_name: The name of the user
        :return: user: the aws_iam.User which got created
        """
        # Validate the users name
        _validate_user_name(user_name)
        id_ = _to_id(user_name)
        if id_ in self.users:
            raise RuntimeError(f"User '{user_name}' (id={id_}) already exists.")

        # make sure the groups exist
        if groups is None:
            resolved_groups = []
        elif isinstance(groups, str):
            resolved_groups = [self.get_group(groups)]
        else:
            resolved_groups = [self.get_group(group) for group in groups]

        # Create the user with some default config and policies
        user = aws_iam.User(
            scope=self,
            # Todo: Not sure what the requirements are and if this is already enough
            id=user_name.lower().replace(".", "-"),
            user_name=user_name,
            password=SecretValue.unsafe_plain_text(INITIAL_PASSWORD),
            password_reset_required=True,
        )

        user.add_managed_policy(self.policy_allow_password_changes)

        for group in resolved_groups:
            user.add_to_group(group)

        self.users[user_name] = user

        return user


# From here it's the actual group and user structure

# Pattern: add groups, then add users with group membership
# 1. Do not attach users to groups, but groups to users
#    (easier off boarding, just delete the add_org_user() call)
# 2. Do not attach policies here, only where you define the service
#    (keep everything service related together)
# TODO: have some though about structuring/naming groups to make them consistent


# constants for groups to make the editor do the hard work...
"""Debugging ability (like reading logs and all data)"""
GROUP_DATA_LAKE_DEBUGGING = "DataLakeDebugging"
"""Run athena queries against the data lake and publish these results in (personal) glue databases"""
GROUP_DATA_LAKE_ATHENA_USER = "DataLakeAthenaUser"


def create_org_groups(scope: Construct):
    """
    Create the OrgUsersAndGroups construct and add all groups

    :param: scope: The parent construct where this should be added to
    :return: the instantiated UsersAndGroups construct
    """
    uag = OrgUsersAndGroups(scope, "org-users-and-groups")
    # Groups

    uag.add_org_group(group_name=GROUP_DATA_LAKE_DEBUGGING)
    uag.add_org_group(group_name=GROUP_DATA_LAKE_ATHENA_USER)
    return uag


def add_users_on_dev(org_groups: OrgUsersAndGroups) -> None:
    """
    Create users which should be present on a dev environment

    :param: org_groups: The already created construct for groups and users
    """
    org_groups.add_org_user(
        "example.user",
        groups=[
            GROUP_DATA_LAKE_DEBUGGING,
            GROUP_DATA_LAKE_ATHENA_USER,
        ],
    )


def add_users_on_prod(org_groups: OrgUsersAndGroups) -> None:
    """
    Create users which should be present on the prod environment

    :param: org_groups: The already created construct for groups and users
    """
    org_groups.add_org_user("example.user", groups=[GROUP_DATA_LAKE_DEBUGGING])
