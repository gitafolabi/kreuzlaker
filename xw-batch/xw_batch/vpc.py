from aws_cdk.aws_ec2 import (
    Vpc,
    SubnetType,
    InterfaceVpcEndpointAwsService,
    GatewayVpcEndpointAwsService,
    GatewayVpcEndpointOptions,
)
from constructs import Construct


class XwVpc(Construct):
    def __init__(self, scope: Construct, id: str, **kwargs) -> None:
        super().__init__(scope, id, **kwargs)

        self.vpc = Vpc(
            self,
            id,
            max_azs=1,
            nat_gateways=0,
            gateway_endpoints={"S3": GatewayVpcEndpointOptions(service=GatewayVpcEndpointAwsService.S3)},
            subnet_configuration=[
                {
                    "cidrMask": 24,
                    "name": "ingress",
                    "subnetType": SubnetType.PUBLIC,
                },
                {
                    "cidrMask": 24,
                    "name": "application",
                    "subnetType": SubnetType.PRIVATE_ISOLATED,
                },
                {
                    "cidrMask": 28,
                    "name": "egress",
                    "subnetType": SubnetType.PRIVATE_WITH_EGRESS,
                },
            ],
        )

        self.vpc.add_interface_endpoint(
            "AthenaEndpoint",
            service=InterfaceVpcEndpointAwsService.ATHENA,
        )

        self.vpc.add_interface_endpoint(
            "EcrEndpoint",
            service=InterfaceVpcEndpointAwsService.ECR,
        )

        self.vpc.add_interface_endpoint(
            "CloudWatchEndpoint",
            service=InterfaceVpcEndpointAwsService.CLOUDWATCH_LOGS,
        )

        self.vpc.add_interface_endpoint(
            "EcrDockerEndpoint",
            service=InterfaceVpcEndpointAwsService.ECR_DOCKER,
        )

    def get_vpc(self):
        return self.vpc
