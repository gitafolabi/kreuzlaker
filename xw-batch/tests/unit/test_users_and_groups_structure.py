import aws_cdk
import pytest
from aws_cdk.assertions import Match, Template

from xw_batch.users_and_groups import OrgUsersAndGroups


class _UaGStack(aws_cdk.Stack):
    def __init__(self):
        super().__init__()
        self.ouag = OrgUsersAndGroups(self, "OrgUsersAndGroups")
        self.ouag.add_org_group("ExampleGroup")
        self.ouag.add_org_user("example.user", groups=["ExampleGroup"])


@pytest.fixture(name="ouag_stack", scope="module")
def ouag_stack_fixture() -> _UaGStack:
    return _UaGStack()


@pytest.fixture(name="ouag_template", scope="module")
def ouag_template_fixture(ouag_stack: _UaGStack) -> Template:
    stack = ouag_stack
    template = Template.from_stack(stack)
    return template


def test_org_users_and_groups_structure(
    ouag_template: Template,
    ouag_stack: _UaGStack,
):
    ouag_template.has_resource_properties(
        "AWS::IAM::Group",
        {
            "GroupName": "ExampleGroup",
        },
    )
    resolved_wanted_policy_arn = ouag_stack.resolve(ouag_stack.ouag.policy_allow_password_changes.managed_policy_arn)
    ouag_template.has_resource_properties(
        "AWS::IAM::User",
        {
            "LoginProfile": {
                "Password": Match.any_value(),
                "PasswordResetRequired": True,
            },
            "ManagedPolicyArns": Match.array_with([resolved_wanted_policy_arn]),
            "UserName": "example.user",
        },
    )

    ouag_template.has_resource_properties(
        "AWS::IAM::ManagedPolicy",
        {
            "Description": "Allow password changed for users.",
            "PolicyDocument": {
                "Statement": [
                    {
                        "Action": "iam:GetAccountPasswordPolicy",
                        "Effect": "Allow",
                        "Resource": "*",
                        "Sid": "ViewAccountPasswordRequirements",
                    },
                    {
                        "Action": ["iam:GetUser", "iam:ChangePassword"],
                        "Effect": "Allow",
                        "Resource": "arn:aws:iam::*:user/${aws:username}",
                        "Sid": "ChangeOwnPassword",
                    },
                ],
            },
        },
    )


def test_org_users_and_groups_snapshot(snapshot, ouag_template):
    assert ouag_template.to_json() == snapshot
