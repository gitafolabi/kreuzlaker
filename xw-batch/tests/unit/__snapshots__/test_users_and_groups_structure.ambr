# name: test_org_users_and_groups_snapshot
  dict({
    'Parameters': dict({
      'BootstrapVersion': dict({
        'Default': '/cdk-bootstrap/hnb659fds/version',
        'Description': 'Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]',
        'Type': 'AWS::SSM::Parameter::Value<String>',
      }),
    }),
    'Resources': dict({
      'OrgUsersAndGroupsdatalakedebuggingE8ABA7E1': dict({
        'Properties': dict({
          'GroupName': 'DataLakeDebugging',
        }),
        'Type': 'AWS::IAM::Group',
      }),
      'OrgUsersAndGroupsexampleuser73FA0DA9': dict({
        'Properties': dict({
          'Groups': list([
            dict({
              'Ref': 'OrgUsersAndGroupsdatalakedebuggingE8ABA7E1',
            }),
          ]),
          'LoginProfile': dict({
            'Password': 'N0tV3ryS3cr3t!',
            'PasswordResetRequired': True,
          }),
          'ManagedPolicyArns': list([
            dict({
              'Ref': 'OrgUsersAndGroupspolicyallowpasswordchanges91D72591',
            }),
          ]),
          'UserName': 'example.user',
        }),
        'Type': 'AWS::IAM::User',
      }),
      'OrgUsersAndGroupspolicyallowpasswordchanges91D72591': dict({
        'Properties': dict({
          'Description': '',
          'ManagedPolicyName': 'AllowPasswordChanges',
          'Path': '/',
          'PolicyDocument': dict({
            'Statement': list([
              dict({
                'Action': 'iam:GetAccountPasswordPolicy',
                'Effect': 'Allow',
                'Resource': '*',
                'Sid': 'ViewAccountPasswordRequirements',
              }),
              dict({
                'Action': list([
                  'iam:GetUser',
                  'iam:ChangePassword',
                ]),
                'Effect': 'Allow',
                'Resource': 'arn:aws:iam::*:user/${aws:username}',
                'Sid': 'ChangeOwnPassword',
              }),
            ]),
            'Version': '2012-10-17',
          }),
        }),
        'Type': 'AWS::IAM::ManagedPolicy',
      }),
    }),
    'Rules': dict({
      'CheckBootstrapVersion': dict({
        'Assertions': list([
          dict({
            'Assert': dict({
              'Fn::Not': list([
                dict({
                  'Fn::Contains': list([
                    list([
                      '1',
                      '2',
                      '3',
                      '4',
                      '5',
                    ]),
                    dict({
                      'Ref': 'BootstrapVersion',
                    }),
                  ]),
                }),
              ]),
            }),
            'AssertDescription': "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
          }),
        ]),
      }),
    }),
  })
# ---
